docs(site) internal page anchors and  referencing mechanism automation

HenadzV opened the pull request:
In the scope of this PR was added automatic functionality for adding headers IDs and creating inpage anchors/referencing links.

Closes:  #1456
Closes:  #1457

HenadzV committed:
feat(docs): Add functions for generating header IDs and anchor links

<body>
This commit adds the `generateHeadersIds` and `generateAnchors` functions to the MDRenderer class. These functions are responsible for generating header IDs and anchor links in the rendered markdown content.

HenadzV committed:
feat(docs): Add function to generate links for globally defined terms

<body>
This commit adds the `generateGloballyDefinedTermsLinks` function to the MDRenderer class. This function generates  links for globally defined terms

ala-n commented on pages/11ty/markdown.shortcut.js:
```diff
@@ -71,6 +77,53 @@ class MDRenderer {
+    });
+  }
+
+  static generateAnchors(content, text, link) {
```
The current process makes the same things (and not the light ones) a couple of times.
Propose to split it a little bit...

We can use the following snippet to find all the text nodes in DOM.
```
function findTextNodes(root: Node): Node[] {
  let all: Node[] = [];
  for (let node = root.firstChild; node; node = node.nextSibling) {
    if (node.nodeType === 3) all.push(node);
    else all = all.concat(findTextNodes(node));
  }
  return all;
}
```

Then we have a simple and straight process:
```
 const nodes = findTextNodes(consent)
   .filter((node) => !node.parentElement.closest('h1, h2, h3, h4, h5, h6'));
 
 for (const [text, link] = Object.entries(terms)) {
    const url = resolveLink(link);
    const regex = new RegExp(`(^|\\s)${text}(\\s|[,\\.])`, 'g');
    for (const node of nodes) {
       node.textContent = node.textContent.replace(regex, `$1<a href="${url}">${text}</a>$2`);
    }
 }
```

ala-n reviewed (changes requested):
Please make sure the approach is optimized and aligned with the project architecture. (according to notes above)

TODO: discuss Mn sync-up
CC: @HenadzV , @abarmina

HenadzV committed:
feat(docs): Reworked links auto generation according pull request remarks

<body>
This commit separates functionality for different methods, adjusts Regex

ala-n commented on pages/11ty/markdown.shortcut.js:
```diff
@@ -71,6 +76,55 @@ class MDRenderer {
+    }
+  }
+
+  static replaceTextNode(document, node, text, link) {
```
```suggestion
  static wrapTextWithLink(document, node, text, link) {
```

HenadzV committed:
feat(docs): Fix variables usage, rename method

<body>
This commit put in order code according project code style guide

HenadzV committed:
feat(docs): fix object assign
<body>
changes according PR review

yadamskaya commented on pages/11ty/markdown.shortcut.js:
```diff
@@ -71,6 +76,56 @@ class MDRenderer {
+      const text = header.textContent;
+      const id = MDRenderer.createIDFromText(text)
+      header.setAttribute('id', id);
+      const anchor = `#${id}`
```
```suggestion
      const anchor = `#${id}`;
```

yadamskaya commented on pages/11ty/markdown.shortcut.js:
```diff
@@ -71,6 +76,56 @@ class MDRenderer {
+      const anchor = `#${id}`
+      localTerms[text] = anchor;
+    }
+    return localTerms
```
```suggestion
    return localTerms;
```

yadamskaya commented on pages/11ty/markdown.shortcut.js:
```diff
@@ -71,6 +76,56 @@ class MDRenderer {
+      if (node.nodeType === 3) all.push(node);
+      else all.push(...MDRenderer.findTextNodes(node));
+    }
+    return all
```
```suggestion
    return all;
```

yadamskaya commented on pages/11ty/markdown.shortcut.js:
```diff
@@ -71,6 +76,56 @@ class MDRenderer {
+
+  static fillReferenceLinks(document, content, terms) {
+    const nodes = MDRenderer.findTextNodes(content)
+      .filter((node) => !node.parentElement.closest('h1, h2, h3, h4, h5, h6'))
```
```suggestion
      .filter((node) => !node.parentElement.closest('h1, h2, h3, h4, h5, h6'));
```

yadamskaya commented on pages/11ty/markdown.shortcut.js:
```diff
@@ -71,6 +76,56 @@ class MDRenderer {
+
+      for (const node of nodes) {
+        if (node.textContent.includes(text)) {
+          MDRenderer.wrapTextNode(document, node, text, link)
```
```suggestion
          MDRenderer.wrapTextNode(document, node, text, link);
```

HenadzV committed:
feat(docs): changes according review
<body>
add missed semicolons

ala-n commented:
@ala-n and @abarmina to check if the actual cases are covered and update related content
